kvs: refactor kvs cache to handle raw data as "primary" data

chu11 opened the pull request:
This PR refactors the KVS cache to favor the storage of raw data over json objects.  This accomplishes several goals.

1) before this PR, KVS cache entries had a data "type" associated with it.  This can lead to a "race" in which how the data was loaded into the KVS cache the first time and could lead to errors/bugs later.  For example, if data was incorrectly loaded into the KVS cache as type json, a later attempt to read the data out of the KVS cache as raw data would fail, leading to an error.

2) This should clean up the code and hopefully make it less confusing.  Many unit tests were removed in this PR, so I feel the less confusing part was accomplished.

The primary idea behind this refactor to remove the "type" system with KVS cache entries and make the KVS cache primarily for storing raw data.

Users can set get/set json objects in the KVS cache, but the API is simply a set of convenience functions converting those json objects to/from their raw string form.

As an aside, I am sometimes anal when it comes to code "ordering".  In commit https://github.com/chu11/flux-core/commit/368f8588bb3b4255914774260e569c3c3f2c43f7  I literally just move "json" code below "raw data" code, b/c I want "raw data" code to be listed first as it is now the "primary" way the KVS cache works.  I know its more code change than may be necessary.

chu11 commented:
One of the builds hung and timed out, restarted it.  One hung and timed out in #1244 earlier today and @garlick suggested one did in #1243.  I'm concerned there is a racy-hang somewhere in the tests.  But it's hard to see where it could be given the output of travis.

garlick commented:
This is a high value cleanup!  I will try provide some review commands before monday.

chu11 commented:
Just missed having coverage on the diff.  Looking through coverage diff, I don't think I can eek out a few more lines.  It's all bad error path scenarios.

garlick commented on src/modules/kvs/cache.c:
```diff
@@ -228,44 +223,58 @@ int cache_entry_set_raw (struct cache_entry *hp, void *data, int len)
+            else if (!hp->data) {
+                /* attempt to change already valid cache entry,
+                 * cannot, must call cache_entry_clear_data() */
+                errno = EBADE;
```
this function doesn't set errno in all the -1 return cases (e.g. if args are invalid)

the second test for !hp->data is redundant with the first leg of the conditional.

maybe it would be better to change the hp->entry_valid leg of the outer conditional to simply
```c
if (hp->entry_valid) {
    if (hp->o)
        json_decref (o); // already stored
    else
        hp->o = o;
    assert (hp->len > 0 && hp->data != NULL); // o/w would not be valid json
} else ...
```

chu11 replied:
ugh, why did I add the redundant check :-)

let me think about how to handle the logic here, as I seemed to have copied some logic from the ```set_raw()``` function, but perhaps it shouldn't have been copied over in that way.

garlick commented on src/modules/kvs/cache.c:
```diff
@@ -228,44 +223,58 @@ int cache_entry_set_raw (struct cache_entry *hp, void *data, int len)
+                json_decref (o);
+            else if (!hp->data) {
+                /* attempt to change already valid cache entry,
+                 * cannot, must call cache_entry_clear_data() */
```
`cache_entry_clear_data()` doesn't appear to have any users

chu11 replied:
Yeah created this function as a "just in case", knowing there were no real use cases at the time.  Perhaps would be wise to remove.  If anything, users can ```cache_remove_entry``` then recreate if they are in dire straits to put new data in the cache entry.

chu11 commented on src/modules/kvs/cache.c:
```diff
@@ -84,42 +84,42 @@ struct cache_entry *cache_entry_create (void)
 }
 
-struct cache_entry *cache_entry_create_raw (void *data, int len)
+struct cache_entry *cache_entry_create_json (json_t *o)
```
This sounds like a good idea.  Was honestly surprised the ```cache_entry_create_raw()``` and ```cache_entry_create_json``` functions had almost no use.

chu11 commented on src/modules/kvs/cache.c:
```diff
@@ -59,8 +59,9 @@ struct cache_entry {
     int len;
-    bool data_valid;          /* flag indicating if data set, don't use
-                             * data == NULL as test */
+    bool entry_valid;       /* flag indicating if data set, don't use
```
agreed that this should be consistent, however I'll make a new issue and clean up in another PR.  Some functions (such as ```cache_entry_clear_dirty()```) return the current setting of the dirty bit as a 1 or 0.  This should be a bool, not a 1 or 0.  It'll be a bit more cleanup than I'd like in this PR.

chu11 commented:
just re-pushed with some fixes.  Removed ```cache_entry_clear_data()```, ```cache_entry_create_raw()```, and ```cache_entry_create_json()```.

Changed the logic in ```cache_entry_set_json()``` a bit to hopefully make more sense.

Removed unnecessary checks for NULL in ```cache_entry_destroy()```

Will do the cleanup regarding consistent dirty bit and valid bit in a separate PR (#1247).

chu11 commented:
coverage diff fell from about 77% to about 71%, I believe mostly due to high coverage functions that were removed.

garlick commented:
IF you're feeling OK with this, I'd say squash it down and we can merge it.

chu11 commented:
squashed and re-pushed
